Feature/arrouse/customizable data source

jkaufman opened the pull request.

alexrrouse committed:
adding a custom datasource

alexrrouse committed:
Updating the requirement of needing to implement CollecitonViewDataSource methods

alexrrouse committed:
Giving support for Revealing cell

alexrrouse committed:
Sloppily blocking horizontal movement

alexrrouse committed:
Adding animation blocks

alexrrouse committed:
Fixing flash in layout

alexrrouse committed:
Fixing the snapshot scale issue on iPhone6

alexrrouse committed:
Putting back in snapshotting but doing it before screen updates

jkaufman commented:
@Raizlabs/maintainers-ios Any comment before I merge?

KingOfBrian commented on LXReorderableCollectionViewFlowLayout/LXReorderableCollectionViewFlowLayout.m:
```diff
@@ -138,7 +140,10 @@ - (void)applyLayoutAttributes:(UICollectionViewLayoutAttributes *)layoutAttribut
+    if ( _dataSource == nil ) {
+        _dataSource = (id<LXReorderableCollectionViewDataSource>)self.collectionView.dataSource;
+    }
+    return _dataSource;
```
I'd have the property declared read only and just return `self.collectionView.dataSource`.   If the collection view changed the data source, this would not be very happy.

spencerp replied:
If the property is `readonly` we won't be able to set the custom data source. Am I understanding your comment correctly?

KingOfBrian replied:
I don't see any code that's setting a custom dataSource.  It's just a convenience accessor into `self.collectionView.dataSource`.   Caching this into our own ivar doesn't make sense to me.

If the property is `readonly`, and we declare the implementation of `-(id)dataSource` without referencing `_dataSource`, the ivar won't be created, via auto-synthesis voodoo.

spencerp replied:
There's no custom data source set within the library, but part of this feature was adding `dataSource` to the public interface. That way any code using this layout can set a custom data source, as opposed to just using the collectionview's datasource, as desired.

For example, in a project in which I'm using this, we are setting a view controller as the layout's data source. That way we can use this in conjunction with `RZCollectionListCollectionViewDataSource`.

KingOfBrian commented on LXReorderableCollectionViewFlowLayout/LXReorderableCollectionViewFlowLayout.m:
```diff
@@ -163,7 +168,7 @@ - (void)invalidateLayoutIfNecessary {
         [self.dataSource collectionView:self.collectionView itemAtIndexPath:previousIndexPath willMoveToIndexPath:newIndexPath];
     }
-
+    
```
These should be removed ideally.   There's an Xcode preferences setting that will cause these not to happen.

KingOfBrian replied:
Would still like to see these go away, minimizing whitespace changes in any PR is a good idea.

spencerp committed:
Removed public read-write access to data source
- Kept dataSource property public so that subclasses can override the getter/setter if necessary

SpencerP commented:
@KingOfBrian as per our offline discussion, I removed the public read-write access to `dataSource` to avoid the data source being modified after instantiation. Property is still public, however, so that subclasses can reimplement if necessary.

KingOfBrian commented on LXReorderableCollectionViewFlowLayout/LXReorderableCollectionViewFlowLayout.m:
```diff
@@ -353,8 +360,13 @@ - (void)handleLongPressGesture:(UILongPressGestureRecognizer *)gestureRecognizer
+                         strongSelf.reorderCompletionAnimationBlock( strongSelf.currentView );
+                         strongSelf.currentView.center = layoutAttributes.center;
+                     }
+                     else {
```
You might want to add an `else if ( strongSelf )` to maintain the original `if (strongSelf)` guard

spencerp replied:
Nice catch.

KingOfBrian commented on LXReorderableCollectionViewFlowLayout/LXReorderableCollectionViewFlowLayout.m:
```diff
@@ -382,7 +394,8 @@ - (void)handlePanGesture:(UIPanGestureRecognizer *)gestureRecognizer {
         case UIGestureRecognizerStateChanged: {
-            self.panTranslationInCollectionView = [gestureRecognizer translationInView:self.collectionView];
+            self.panTranslationInCollectionView = CGPointMake(0.0f, [gestureRecognizer translationInView:self.collectionView].y);
+            
```
Disabling `.x` panning seems to be a substantial change.   I'm guessing this is desired for your usage, but may break others.   Maybe add a `BOOL disableHorizontalPanning`?

KingOfBrian commented on LXReorderableCollectionViewFlowLayout/LXReorderableCollectionViewFlowLayout.m:
```diff
@@ -299,7 +301,7 @@ - (void)handleLongPressGesture:(UILongPressGestureRecognizer *)gestureRecognizer
             UIView *imageView = [collectionViewCell LX_snapshotView];
             imageView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
-            imageView.alpha = 0.0f;
+            imageView.alpha = 1.0f;
```
This also seems like a substantial behavior change for other use cases.   Maybe make it conditional on the presence of `reorderStartAnimationBlock`

spencerp committed:
Removing whitespace changes as per code review

KingOfBrian commented on LXReorderableCollectionViewFlowLayout/LXReorderableCollectionViewFlowLayout.m:
```diff
@@ -258,7 +260,7 @@ - (void)handleScroll:(CADisplayLink *)displayLink {
             }
             
-            translation = CGPointMake(distance, 0.0f);
+            translation = CGPointMake(0.0f, 0.0f);
```
The above two may be a case where the horizontal scrolling being disabled should be bool backed.

spencerp replied:
These are actually made unnecessary by the restriction on `panTranslationInCollectionView` in `handlePanGesture`. I'll remove the changes.

KingOfBrian commented on LXReorderableCollectionViewFlowLayout/LXReorderableCollectionViewFlowLayout.m:
```diff
@@ -50,8 +50,9 @@ - (UIView *)LX_snapshotView;
+    
     if ([self respondsToSelector:@selector(snapshotViewAfterScreenUpdates:)]) {
-        return [self snapshotViewAfterScreenUpdates:YES];
+        return [self snapshotViewAfterScreenUpdates:NO];
```
What's the reason for this out of curiosity?

spencerp replied:
Seems that it has to do with a bug on iPhone 6/6+'s when zooming the display (running an app that is not optimized for the larger screen size).

http://stackoverflow.com/q/25873234/1959707

spencerp committed:
Code cleanup
- Made highlighted cell snapshot contingent upon cell’s state at pickup
- Added ability to limit view drag direction based on scroll direction
- Misc code hygiene

KingOfBrian commented on LXReorderableCollectionViewFlowLayout/LXReorderableCollectionViewFlowLayout.m:
```diff
@@ -291,18 +292,30 @@ - (void)handleLongPressGesture:(UILongPressGestureRecognizer *)gestureRecognizer
+            
+            // Match snapshot state to state of cell when it was lifted
+            if ( cellHighlightedWhenPickedUp ) {
+                collectionViewCell.highlighted = YES;
```
It's already `YES` if we get here.

KingOfBrian commented on LXReorderableCollectionViewFlowLayout/LXReorderableCollectionViewFlowLayout.m:
```diff
@@ -291,18 +292,30 @@ - (void)handleLongPressGesture:(UILongPressGestureRecognizer *)gestureRecognizer
+                imageView.alpha = 0.0f;
+            }
+            else {
+                highlightedImageView.alpha = 0.0f;
```
This guy is `nil` here.

spencerp replied:
I guess we don't need this `else` at all, actually.

spencerp committed:
Removed redundant logic in handleLongPressGesture:

KingOfBrian commented on LXReorderableCollectionViewFlowLayout/LXReorderableCollectionViewFlowLayout.m:
```diff
@@ -291,18 +292,25 @@ - (void)handleLongPressGesture:(UILongPressGestureRecognizer *)gestureRecognizer
-            highlightedImageView.alpha = 1.0f;
+            BOOL cellHighlightedWhenPickedUp = collectionViewCell.highlighted;
+            UIView *highlightedImageView;
+            UIView *imageView;
```
These two guys can move back inline and it should be a bit cleaner, `highlightedImageView` is only used in the conditional.

spencerp committed:
Code cleanup

jkaufman commented:
:+1:
